numpyro/infer/ensemble.py

```diff
@@ -0,0 +1,735 @@
+            z_flat = jax.lax.cond(i == 0, 
+                                  lambda x: x.at[:split_ind].set(z_updates),
+                                  lambda x: x.at[split_ind:].set(z_updates),
+                                  z_flat)
+            return (z_flat, inner_state)
+                
+        z_flat, inner_state = jax.lax.fori_loop(0, 2, body_fn, (z_flat, inner_state))
+
+
+        return EnsembleSamplerState(unravel_fn(z_flat), inner_state, rng_key)
+
+
+class AIES(EnsembleSampler):
+    """
+    Affine-Invariant Ensemble Sampling: a gradient free method. Suitable for low to moderate dimensional models.
+    Generally, `num_chains` should be at least twice the dimensionality of the model.
+    
+    .. note:: This kernel must be used with `num_chains` > 1 and `chain_method="vectorized`
+        or `chain_method="parallel` in :class:`MCMC`. The number of chains must be divisible by 2.
+    
+    **References:**
+    
+    1. *emcee: The MCMC Hammer* (https://iopscience.iop.org/article/10.1086/670067),
+        Daniel Foreman-Mackey, David W. Hogg, Dustin Lang, and Jonathan Goodman. 
+    
+    :param model: Python callable containing Pyro :mod:`~numpyro.primitives`.
+        If model is provided, `potential_fn` will be inferred using the model.
+    :param potential_fn: XXX currently unsupported.
+    :param bool randomize_split: whether or not to permute the chain order at each iteration.
+    :param moves: a dictionary mapping moves to their respective probabilities of being selected.
+        If left empty, defaults to `AIES.DEMove()`. 
+    :param callable init_strategy: a per-site initialization function.
+        See :ref:`init_strategy` section for available functions.
+        
+    **Example**
+
+    .. code-block:: python
+        import jax
+        import jax.numpy as jnp
+        import numpyro
+        import numpyro.distributions as dist
+        from numpyro.infer import MCMC, AIES
+
+        def model():
+            x = numpyro.sample("x", dist.Normal().expand([10]))
+            numpyro.sample("obs", dist.Normal(x, 1.0), obs=jnp.ones(10))
+        
+        kernel = AIES(model, moves={AIES.DEMove() : .5, 
+                                    AIES.StretchMove() : .5})
+        mcmc = MCMC(kernel, num_warmup=1000, num_samples=2000, num_chains=20, chain_method='vectorized')
+        mcmc.run(jax.random.PRNGKey(0))
+        mcmc.print_summary()
+    """
+    
+    def __init__(self, model=None, potential_fn=None, randomize_split=False, moves=None, init_strategy=init_to_uniform):
+        if not moves:
+            self._moves = [AIES.DEMove()]
+            self._weights = jnp.array([1.0])
+        else:
+            self._moves = list(moves.keys())
+            self._weights = jnp.array([weight for weight in moves.values()]) / len(moves)
+
+        super().__init__(model, potential_fn, randomize_split, init_strategy)
+
+    # XXX: this doesn't show because state_method='vectorized' shuts off diagnostics_str
```

Comment:
Is this `chain_method` @amifalk? I don't think progress_bar works with the vectorized method there.

Reply:
Yes it is, sorry for the typo. Can we adjust [this](https://github.com/pyro-ppl/numpyro/blob/3ef0423ae96d3b21bc0bc09d497fc772e7db8dc0/numpyro/infer/mcmc.py#L427 ) line to allow the diagnostics_str for these ensemble methods? If I remove the prng_key check, it displays correctly.

Reply:
That line checks whether `vectorized` method is used. If rng_key.ndim == 1, parallel is used. Otherwise, vectorized is used. Looking at your code, it seems that rng_key.ndim == 2 in both cases, which is strange to me. Could you double check the logic for both methods?

I think you can add an attribute to MCMCKernel to indicate whether it is an ensemble kernel and skip `is_prng_key` check at that line if the kernel is an ensemble one.

Reply:
Good catch. Parallel sampling does not work with pmap since the chains need to talk to each other. To make that work, we would probably need to add an internal utility that does parallel sharding on the chains, kind of like what is requested in #1192. In the interest of not overcomplicating the API, I'll opt not to support it for now and make a note in the code, but if there was sufficient interest I think we could add a `shard_chains` argument to `EnsembleSampler`.